test/image-tests/ostree.sh

```diff
@@ -0,0 +1,479 @@
+
+bootloader --timeout=1 --append="net.ifnames=0 modprobe.blacklist=vc4"
+
+network --bootproto=dhcp --device=link --activate --onboot=on
+
+zerombr
+clearpart --all --initlabel --disklabel=msdos
+autopart --nohome --noswap --type=plain
+ostreesetup --nogpg --osname=${IMAGE_TYPE} --remote=${IMAGE_TYPE} --url=http://192.168.100.1/repo/ --ref=${OSTREE_REF}
+poweroff
+
+%post --log=/var/log/anaconda/post-install.log --erroronfail
+
+# no sudo password for user admin
+echo -e 'admin\tALL=(ALL)\tNOPASSWD: ALL' >> /etc/sudoers
+
+# Remove any persistent NIC rules generated by udev
+rm -vf /etc/udev/rules.d/*persistent-net*.rules
+# And ensure that we will do DHCP on eth0 on startup
+cat > /etc/sysconfig/network-scripts/ifcfg-eth0 << EOF
+DEVICE="eth0"
+BOOTPROTO="dhcp"
+ONBOOT="yes"
+TYPE="Ethernet"
+PERSISTENT_DHCLIENT="yes"
+EOF
+
+echo "Packages within this iot or edge image:"
+echo "-----------------------------------------------------------------------"
+rpm -qa | sort
+echo "-----------------------------------------------------------------------"
+# Note that running rpm recreates the rpm db files which aren't needed/wanted
+rm -f /var/lib/rpm/__db*
+
+echo "Zeroing out empty space."
+# This forces the filesystem to reclaim space from deleted files
+dd bs=1M if=/dev/zero of=/var/tmp/zeros || :
+rm -f /var/tmp/zeros
+echo "(Don't worry -- that out-of-space error was expected.)"
+
+%end
+STOPHERE
+
+# Install ostree image via anaconda.
+greenprint "Install ostree image via anaconda"
+sudo virt-install  --initrd-inject=${KS_FILE} \
+                   --extra-args="ks=file:/ks.cfg console=ttyS0,115200" \
+                   --name=${IMAGE_KEY}\
+                   --disk path=${LIBVIRT_IMAGE_PATH},format=qcow2 \
+                   --ram 3072 \
+                   --vcpus 2 \
+                   --network network=integration,mac=34:49:22:B0:83:30 \
+                   --os-type linux \
+                   --os-variant ${OS_VARIANT} \
+                   --location ${BOOT_LOCATION} \
+                   --nographics \
+                   --noautoconsole \
+                   --wait=-1 \
+                   --noreboot
+
+# Start VM.
+greenprint "Start VM"
+sudo virsh start ${IMAGE_KEY}
+
+# Check for ssh ready to go.
+greenprint "🛃 Checking for SSH is ready to go"
+for LOOP_COUNTER in `seq 0 30`; do
+    RESULTS="$(wait_for_ssh_up $GUEST_ADDRESS)"
+    if [[ $RESULTS == 1 ]]; then
+        echo "SSH is ready now! 🥳"
+        break
+    fi
+    sleep 10
+done
+
+# Check image installation result
+check_result
+
+##################################################
+##
+## ostree image/commit upgrade
+##
+##################################################
```

Comment:
At first glance I am not sure what the script below does. I see building 2 images with composer-cli but it is unclear to me why do we have to extract them and then pull/commit with osbuild ? Can you explain what the usage scenario here is ?

Reply:
The first one is for image installation test, and the second one is for ostree upgrade test. So you have to give it `parent commit` and ostree `ref` args in composer-cli.
To run ostree upgrade test, the ostree repo needs a new commit which made some change, add wget package in this case. So a new image has to be built with ostree `parent commit` and `ref` configured, they're two args of composer-cli.
Then use local pull to `merge` upgrade ostree commit into installation commit and summary the change.
After that the ostree `client` is able to do upgrade.

Reply:
OK, I see what the intention here is but that looks like a very complicated use case to me. 
1) Build an image with composer-cli
2) Extract it under /var/www/html
3) Then use anaconda + kickstart to create a VM from the extracted contents

4) Repeat 1-3 only so that we can try an upgrade procedure. 
5) Call the ansible playbook to verify the internals for the running VM


1-3 somewhat duplicate osbuild-image-tests but not entirely. I've opened #962 to start cleaning this up. 

4) looks like a good use-case but in case of problems I have to ask what the root cause of that problem would be ? On line 440 `rpm-ostree upgrade` could possibly fail or maybe fail on 427-428 when pulling the new commits into the ostree repository.  But all of that would indicate a faulty image coming out of osbuild-composer (assuming extraction works fine). Can't this be verified with some sort of a content sanity test? E.g. isn't the "image info" portion of osbuild-image-tests ensuring exactly that ? 

I can see the e2e path here but it really looks like we're testing `osbuild pull-local/update` and `rpm-ostree upgrade` instead of Image Builder.

for 5) shouldn't the script call the ansible playbook somewhere around line 380, e.g. after the initial VM is booted up; That will examine the starting state of the VM before upgrade and looks like if you pass the appropriate values from the outside the playbook will work fine.

Reply:
Thanks @atodorov. :+1: 
1. Issue happened between 3 and 4. That means VM installed but can't boot, at least can't ssh to VM.
2. I have no idea about how to do ostree content sanity test. There's no issue or error in build log.
3. I run the same code in another rhel 8.3 VM, everything works well.
4. Adding ansible playbook after ostree installation is a good idea. I'll update that.

Reply:
> I can see the e2e path here but it really looks like we're testing osbuild pull-local/update and rpm-ostree upgrade instead of Image Builder.

Just to answer to this: basically the properties we need from the commits is that they can be installed and upgraded (and more things, but those are the basics). There is no way that I know to verify this without going through the motions of the end-to-end test. One could argue that some of this should be tested elsewhere, but as far as I know that is not being done so we are sort of stuck with having to verify that.
I do agree that tests like image-info /should/ catch most things, but that depends on us having gotten the `image-info` tool correct, so I think doing the e2e is the only way to make sure.